ui/dashboards/src/context/LayoutsSlice.tsx

```diff
@@ -0,0 +1,49 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { LayoutDefinition, GridItemDefinition } from '@perses-dev/core';
+import produce from 'immer';
+import { StateCreator } from 'zustand';
+import { useDashboardStore } from './DashboardProvider';
+
+export interface LayoutsSlice {
+  updateLayout: (layout: LayoutDefinition, index?: number) => void;
+  addItemToLayout: (index: number, item: GridItemDefinition) => void;
+}
+
+export const createLayoutsSlice: StateCreator<LayoutsSlice> = (set) => ({
+  updateLayout: (layout: LayoutDefinition, index?: number) =>
+    set(
+      produce((state) => {
```

Comment:
You can also use the immer-middleware so you don't have to do this in each set:
https://github.com/pmndrs/zustand#immer-middleware

Reply:
i need access to `layouts` in this slice, but i'm not sure how i can pass `layouts` or 'props` using the immer middleware since i need to initialize `layouts` from props

any idea?

Reply:
I think you can use `get()`? 
https://github.com/pmndrs/zustand#read-from-state-in-actions

Reply:
ok i finally figured it out - i just had to cast the `state` type to `DashboardStoreState`

Reply:
get() is read only